### PR TITLE
Remove pod mutation for PVs with supplemental GIDs

### PR DESCRIPTION
**Pull request by wongma7**
Continuation of https://github.com/kubernetes/kubernetes/pull/27571 to bring the feature back without pod mutation

### CODE REVIEW

**Comment by jsafrane** [comment]
@k8s-bot ok to test

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 1]
```diff
@@ -392,32 +381,27 @@ func getExpectedVolumes(pod *api.Pod) []string {
...
+	if volumeGidValue == "" {
+		return 0, false
+	}
 
-		if pod.Spec.SecurityContext == nil {
-			pod.Spec.SecurityContext = &api.PodSecurityContext{}
-		}
+	gid, err := strconv.ParseInt(volumeGidValue, 10, 64)
+	if err != nil {
+		// TODO: should be unreachable code due to validations
```
Comment (jsafrane) [inline pkg/kubelet/volumemanager/volume_manager.go]: I would remove TODO from this comment - there is nothing to do here.

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+	expected := pod.Spec.Volumes[0].Name
+
+	manager, err := newTestVolumeManager(tmpDir, podManager, kubeClient)
+	if err != nil {
+		t.Fatalf("Failed to initialize volume manager: %v", err)
+	}
+
+	stopCh := make(chan struct{})
+	go manager.Run(stopCh)
+	defer func() {
```
Comment (jsafrane) [inline pkg/kubelet/volumemanager/volume_manager_test.go]: you can directly `defer close(stopCh)`

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+	expected := []api.UniqueVolumeName{api.UniqueVolumeName(node.Status.VolumesAttached[0].Name)}
+
+	manager, err := newTestVolumeManager(tmpDir, podManager, kubeClient)
+	if err != nil {
+		t.Fatalf("Failed to initialize volume manager: %v", err)
+	}
+
+	stopCh := make(chan struct{})
+	go manager.Run(stopCh)
+	defer func() {
```
Comment (jsafrane) [inline pkg/kubelet/volumemanager/volume_manager_test.go]: `defer close(stopCh)`

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+		actual := manager.GetExtraSupplementalGroupsForPod(pod)
+		if !reflect.DeepEqual(tc.expected, actual) {
+			t.Errorf("Expected supplemental groups %v, got %v", tc.expected, actual)
+		}
+
+		close(stopCh)
+	}
+}
+
+func TestGetVolumesInUse(t *testing.T) {
```
Comment (jsafrane) [inline pkg/kubelet/volumemanager/volume_manager_test.go]: This is very similar to TestGetMountedVolumesForPod, can it be merged together? I don't like this copy-paste style.

**Comment by jsafrane** [comment]
almost LGTM, just few minor issues. And broken unit test.

**Comment by jsafrane** [comment]
@k8s-bot test this issue: #22655

**Comment by jsafrane** [comment]
Just FYI, some tests did not compile:
```
# k8s.io/kubernetes/pkg/kubelet/dockertools
./docker_manager_test.go:50: imported and not used: "k8s.io/kubernetes/pkg/kubelet/volumemanager"
./docker_manager_test.go:130: undefined: volume in volume.NewFakeVolumeManager
godep: go exit status 2
```

**Comment by wongma7** [comment]
merged the 2 tests and added back 'fake_docker_manager.go' to fix the broken test

**Comment by childsb** [comment]
@k8s-bot test this issue #26760

File: pkg/kubelet/volumemanager/volume_manager.go [hunk 5]
```diff
@@ -98,16 +98,10 @@ type VolumeManager interface {
...
-	// are successfully attached and mounted. The key in the map is the
-	// OuterVolumeSpecName (i.e. pod.Spec.Volumes[x].Name).
-	// It returns an empty VolumeMap if pod has no volumes.
-	// In addition for every volume that specifies a VolumeGidValue, it appends
-	// the SecurityContext.SupplementalGroups for the specified pod.
-	// XXX: https://github.com/kubernetes/kubernetes/issues/27197 mutating the
-	// pod object is bad, and should be avoided.
-	GetVolumesForPodAndAppendSupplementalGroups(pod *api.Pod) container.VolumeMap
+	// GetExtraSupplementalGroupsForPod returns a list of the extra
+	// supplemental groups for the Pod.  These extra supplemental grops come
```
Comment (jsafrane) [inline pkg/kubelet/volumemanager/volume_manager.go]: typo: grops -> groups

**Comment by jsafrane** [comment]
@k8s-bot test this issue #IGNORE

**Comment by jsafrane** [comment]
There is something wrong with this PR that prevents installation on GCE. I don't have time to debug this, but kube-up.sh with this PR fails, while it succeeds without it.

**Comment by jsafrane** [comment]
from kubelet.log:

```
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: I0713 16:50:10.240002    1499 kubelet.go:2642] Recording NodeHasSufficientMemory event message for node jsafrane-e2e-master
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: I0713 16:50:10.285561    1499 kube_docker_client.go:295] Stop pulling image "gcr.io/google_containers/pause-amd64:3.0": "Status: Downloaded newer i
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: E0713 16:50:10.288347    1499 runtime.go:58] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memo
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:52
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:40
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /usr/local/go/src/runtime/asm_amd64.s:472
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /usr/local/go/src/runtime/panic.go:443
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /usr/local/go/src/runtime/panic.go:62
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /usr/local/go/src/runtime/sigpanic_unix.go:24
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/docker_manager.go:699
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/docker_manager.go:1544
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/docker_manager.go:1729
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/dockertools/docker_manager.go:1995
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1815
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:520
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:171
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:154
Jul 13 16:50:10 jsafrane-e2e-master kubelet[1499]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:215
```

**Comment by lavalamp** [comment]
This seems to have broken a bunch of tests in the slow suite.

[k8s.io] Pod Disks should schedule a pod w/ a readonly PD on two hosts, then remove both ungracefully. [Slow]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pd.go:243 Expected error: <*errors.errorString | 0xc8200e2100>: { s: "timed out waiting for the condition", } timed out waiting for the condition not to have occurred


[k8s.io] Pod Disks Should schedule a pod w/ a readonly PD on two hosts, then remove both gracefully. [Slow]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pd.go:291 Expected error: <*errors.errorString | 0xc82007bfc0>: { s: "timed out waiting for the condition", } timed out waiting for the condition not to have occurred


[k8s.io] Pod Disks should schedule a pod w/ a RW PD, ungracefully remove it, then schedule it on another host [Slow]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pd.go:135 Expected error: <*errors.errorString | 0xc8200d6100>: { s: "timed out waiting for the condition", } timed out waiting for the condition not to have occurred


[k8s.io] Pod Disks Should schedule a pod w/ a RW PD, gracefully remove it, then schedule it on another host [Slow]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pd.go:195 Expected error: <*errors.errorString | 0xc8200e0100>: { s: "timed out waiting for the condition", } timed out waiting for the condition not to have occurred

**Pull request merged**